<?xml version="1.0"?>
<launch>
  <!-- Calibration settings -->
  <arg name="namespace_prefix" default="my_eih_calib"/>
  <arg name="marker_id" default="582"/>
  <arg name="marker_size" default="0.083"/>
  <arg name="corner_refinement" default="LINES"/>
  
  <!-- Robot configuration -->
  <arg name="robot_name" default="iiwa"/>
  <arg name="model" default="iiwa7"/>
  <arg name="hardware_interface" default="PositionJointInterface"/>
  <arg name="robot_base_frame" default="iiwa_link_0"/>
  <arg name="robot_effector_frame" default="iiwa_link_ee"/>
  
  <!-- Camera configuration -->
  <arg name="tracking_base_frame" default="camera_base"/>
  <arg name="camera_frame" default="rgb_camera_link"/>
  <arg name="marker_frame" default="aruco_marker_frame"/>
  <arg name="camera_image_topic" default="/rgb/image_raw"/>
  <arg name="camera_info_topic" default="/rgb/camera_info"/>
  
  <!-- Azure Kinect settings -->
  <arg name="depth_mode" value="NFOV_2X2BINNED"/>
  <arg name="color_resolution" value="720P"/>
  <arg name="fps" value="5"/>
  <arg name="rgb_point_cloud" value="false"/>
  <arg name="point_cloud" value="false"/>
  
  <!-- Movement settings -->
  <arg name="freehand_robot_movement" default="false"/>
  
  <!-- Visualization -->
  <arg name="backend_rviz" default="true"/>
  <arg name="world_frame" default="world"/>
  <arg name="publish_world_tf" default="true"/>
  <arg name="world_to_robot_xyz" default="0 0 0"/>
  <arg name="world_to_robot_rpy" default="0 0 0"/>

  <!-- Start robot + MoveIt -->
  <include file="$(find iiwa_moveit)/launch/moveit_planning_execution.launch">
    <arg name="sim" value="false"/>
    <arg name="rviz" value="$(arg backend_rviz)"/>
    <arg name="hardware_interface" value="$(arg hardware_interface)"/>
    <arg name="robot_name" value="$(arg robot_name)"/>
    <arg name="model" value="$(arg model)"/>
  </include>

  <!-- World frame publisher -->
  <group if="$(arg publish_world_tf)">
    <node name="world_to_$(arg robot_base_frame)" pkg="tf" type="static_transform_publisher"
          args="$(arg world_to_robot_xyz) $(arg world_to_robot_rpy) $(arg world_frame) $(arg robot_base_frame) 30"/>
  </group>

  <!-- Azure Kinect camera -->
  <include file="$(find azure_kinect_ros_driver)/launch/driver.launch">
    <arg name="depth_mode" value="$(arg depth_mode)"/>
    <arg name="color_resolution" value="$(arg color_resolution)"/>
    <arg name="fps" value="$(arg fps)"/>
    <arg name="point_cloud" value="$(arg point_cloud)"/>
    <arg name="rgb_point_cloud" value="$(arg rgb_point_cloud)"/>
  </include>

  <!-- ArUco marker tracking -->
  <node pkg="aruco_ros" type="single" name="aruco_single">
    <remap from="image" to="$(arg camera_image_topic)"/>
    <remap from="camera_info" to="$(arg camera_info_topic)"/>
    <param name="image_is_rectified" value="true"/>
    <param name="marker_size" value="$(arg marker_size)"/>
    <param name="marker_id" value="$(arg marker_id)"/>
    <param name="reference_frame" value="$(arg tracking_base_frame)"/>
    <param name="camera_frame" value="$(arg camera_frame)"/>
    <param name="marker_frame" value="$(arg marker_frame)"/>
    <param name="corner_refinement" value="$(arg corner_refinement)"/>
  </node>
  
  <node pkg="easy_handeye" type="pose_to_tf.py" name="aruco_pose_to_tf" output="screen">
    <param name="pose_topic" value="/aruco_single/pose"/>
    <param name="parent_frame" value="camera_base"/>
    <param name="child_frame" value="aruco_marker_frame"/>
  </node>




  <!-- Hand-eye calibration -->
  <include file="$(find easy_handeye)/launch/calibrate.launch">
    <arg name="eye_on_hand" value="true"/>
    <arg name="namespace_prefix" value="$(arg namespace_prefix)"/>
    <arg name="move_group_namespace" value="$(eval '/%s' % arg('robot_name'))"/>
    <arg name="robot_base_frame" value="$(arg robot_base_frame)"/>
    <arg name="robot_effector_frame" value="$(arg robot_effector_frame)"/>
    <arg name="tracking_base_frame" value="$(arg tracking_base_frame)"/>
    <arg name="tracking_marker_frame" value="$(arg marker_frame)"/>
    <arg name="freehand_robot_movement" value="$(arg freehand_robot_movement)"/>
    <arg name="start_rviz" value="false"/>
  </include>
</launch>
